modules/programs/firefox.nix

```diff
@@ -444,6 +532,124 @@ in {
+                        (optionalAttrs (config ? iconUpdateURL) {
+                          # Convenience to default iconURL to iconUpdateURL so
+                          # the icon is immediately downloaded from the URL
+                          iconURL = config.iconURL or config.iconUpdateURL;
+                        } // optionalAttrs (config ? icon) {
+                          # Convenience to specify absolute path to icon
+                          iconURL = "file://${config.icon}";
+                        } // {
+                          # Required for custom engine configurations,
+                          # loadPaths are unique identifiers that are generally
+                          # formatted like: [source]/path/to/engine.xml
+                          loadPath = ''
+                            [home-manager]/programs.firefox.profiles.${profile.name}.search.engines."${
+                              replaceChars [ "\\" ] [ "\\\\" ] name
+                            }"'';
+                        })) // {
+                          # Required fields for all engine configurations
+                          inherit name isAppProvided metaData;
+                        })) orderedEngineList;
+                in engines;
+
+                metaData = optionalAttrs (profile.search.default != null) {
+                  current = profile.search.default;
+
+                  # Hash algorithm from
+                  # https://searchfox.org/mozilla-central/rev/de15f9c109f9c474d00faf8032f559c236067c06/toolkit/components/search/SearchUtils.jsm#312-334
+                  hash = let
+                    # home-manager doesn't circumvent user consent and isn't
+                    # acting maliciously. We're modifying the search outside of
+                    # Firefox, but a claim by Mozilla to remove this would be very
+                    # anti-user, and is unlikely to be an issue for our use case.
+                    disclaimer =
+                      "By modifying this file, I agree that I am doing so "
+                      + "only within $appName itself, using official, user-driven search "
+                      + "engine selection processes, and in a way which does not circumvent "
+                      + "user consent. I acknowledge that any attempt to change this file "
+                      + "from outside of $appName is a malicious act, and will be responded "
+                      + "to accordingly.";
+
+                    salt = profile.path + profile.search.default
+                      + (builtins.replaceStrings [ "$appName" ] [ "Firefox" ]
+                        disclaimer);
+
+                    hash = removeSuffix "\n" (builtins.readFile
+                      (pkgs.runCommand "hash" { inherit salt; } ''
+                        echo -n "$salt" | ${pkgs.openssl}/bin/openssl dgst -sha256 -binary | base64 > "$out"
+                      ''));
+                  in hash;
```

Comment:
This is IFD. We can hash strings within Nix though and perhaps it's not too hard to do base64 encoding?

Why do we need the value of the hash in Nix though? Isn't it always the same?

Couldn't we just substitute it in the JSON? This way we could avoid the IFD.

Reply:
Oh right that's a great point This definitely shouldn't be an IFD. I think I just did it this way because the hashing defaults to base16.

I'll try looking into a Nix-only alternative.

Reply:
Looking into this further, I don't think this is easily doable in pure Nix. You'd need to implement a base16 to base64 converter.

See https://github.com/NixOS/nix/issues/3151

I think the substitution approach is the best we can do.